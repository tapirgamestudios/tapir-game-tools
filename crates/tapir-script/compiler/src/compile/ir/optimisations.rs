use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
