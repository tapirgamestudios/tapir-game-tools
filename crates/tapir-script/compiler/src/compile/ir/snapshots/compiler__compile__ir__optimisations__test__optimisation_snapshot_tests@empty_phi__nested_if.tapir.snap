---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/empty_phi/nested_if.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 3 ----
x.5 = bool true
y.6 = bool true
temp.7 = x.5
conditional_jump temp.7 0 1

---- block 0 ----
temp.8 = y.6
conditional_jump temp.8 4 5

---- block 1 ----
jmp 2

---- block 4 ----
jmp 6

---- block 5 ----
jmp 6

---- block 2 ----
x.9 = φ(x.5 from 1, x.10 from 6)
z.11 = x.9
return 

---- block 6 ----
x.10 = φ(x.5 from 4, x.5 from 5)
jmp 2


----------- optimised -------------
=======================
fn @toplevel() -> 

---- block 3 ----
x.5 = bool true
y.6 = bool true
temp.7 = x.5
conditional_jump temp.7 0 1

---- block 0 ----
temp.8 = y.6
conditional_jump temp.8 4 5

---- block 1 ----
jmp 2

---- block 4 ----
jmp 6

---- block 5 ----
jmp 6

---- block 2 ----
z.11 = x.5
return 

---- block 6 ----
jmp 2
